policy: include L3-only policy in L4Filter

ianvernon opened the pull request:
This allows for all policy at L3 to be stored in a single construct, and means that all resolved policy is now specified in terms of selectors.
    
Resolution to identities is now done in the translation to PolicyMap keys and in the xDS layer, apart from when requirements are analyzed. Because L3-only policy is computed as part of L4Filter, and we still need to calculate the sets of denied identities for xDS policy, we only need to calculate whether requirements deny specific identities now in `ResolvePolicy`. Do just that, and factor out the code which calculates requirements, as well as analysis of `ToEndpoints` and `FromEndpoints` into separate functions to be reused elsewhere, specifically the policy tracing logic.

This allows us to potentially change how tracing works, since we store computed policy in the L4Filter for L3 as well now, but I've kept it the same for now to reduce the amount of changes in a single PR.

Fixes: #7516 

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7536)
<!-- Reviewable:end -->

Commit:
policy: store L3-only policy in L4Filter

This allow for all policy at L3 to be stored in a single construct, and
means that all resolved policy is now specified in terms of selectors.

Resolution to identities is now done in the translation to PolicyMap keys
and in the xDS layer.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: factor out calculation of ingress requirements / label-based L3 into separate functions

Because L3-only policy is computed as part of L4Filter, and we still need to
calculate the sets of denied identities for xDS policy, we only need to
calculate whether requirements deny specific identities now in `ResolvePolicy`.
Do just that, and factor out the code which calculates requirements, as well as
analysis of `FromEndpoints` into separate functions to be reused elsewhere,
specifically the policy tracing logic.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: factor out calculation of egress requirements / label-based L3 into separate functions

Because L3-only policy is computed as part of L4Filter, and we still need to
calculate the sets of denied identities for xDS policy, we only need to
calculate whether requirements deny specific identities now in `ResolvePolicy`.
Do just that, and factor out the code which calculates requirements, as well as
analysis of `ToEndpoints` into separate functions to be reused elsewhere,
specifically the policy tracing logic.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test: specify which container is trying to access world

This makes it easier to debug the egress test fails.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: do not create wildcard at L3 PolicyMap Key for L3-only keys

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/policy/api/selector.go:
```diff
@@ -362,11 +362,17 @@ func (s EndpointSelectorSlice) Matches(ctx labels.LabelArray) bool {
+// behavior: at L3, an empty slice does not allow traffic, while at L4, it does.
+// If a slice contains the WildcardEndpointSelector, it of course selects all
+// endpoints.
+func (s EndpointSelectorSlice) SelectsAllEndpoints(isL3Only bool) bool {
```
This makes it nice and explicit, but we have to pass around this bool all over the place which is ugly. I'm OK with this as an intermediate step but IMO what we should really do is canonicalize the `EndpointSelectorSlice`. Early on, if parsing an L4 rule we find that the slice is empty, add the wildcard selector into it. Then beyond that point, all code can assume that the contents of the `EndpointSelectorSlice` means exactly what its contents indicate --- empty is nothing. Containing selectors means things are selected. No special casing.

ianvernon replied:
Done.

joestringer commented on pkg/policy/rule.go:
```diff
@@ -455,6 +455,48 @@ func (r *rule) canReachIngress(ctx *SearchContext, state *traceState) api.Decisi
+	return api.Undecided
+}
+
+// canReachIngress returns the decision as to whether the set of labels specified
```
This comment needs updating. (I think I found another case like this somewhere as well)

ianvernon replied:
Done.

joestringer commented on pkg/policy/rule.go:
```diff
@@ -455,6 +455,48 @@ func (r *rule) canReachIngress(ctx *SearchContext, state *traceState) api.Decisi
 	return api.Undecided
 }
 
+func (r *rule) canReachIngressV2(ctx *SearchContext, state *traceState) api.Decision {
```
These functions are just copied fragments of `canReachIngress()`, right? I presume that you avoided removing `canReachIngress()` so that you don't have to go and update all the tests. Could you rework `canReachIngress()` to call these functions? I think that would make the patch diff much easier, to show that there are no functional changes here. Otherwise doing a diff between the existing version and this new version is a bit tricky.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -144,7 +144,7 @@ func (l4 *L4Filter) ToKeys(direction trafficdirection.TrafficDirection, identity
 	proto := uint8(l4.U8Proto)
 
-	if l4.AllowsAllAtL3() {
+	if l4.AllowsAllAtL3() && l4.Port != 0 {
```
If L4.Port is 0 and it allows all at L3, won't this generate BPF entries with (0, 0)? The datapath won't respect entries like that right now. Either we need to change the datapath to support this or we should avoid making this change, so that we still go through all the identities and generate the L3 allows for all peer identities.

joestringer replied:
With a bit more thinking here, it's not that bad if we generate this entry so long as it doesn't shortcut the generation of entries with non-zero identity + zero port (which I suspect it may end up doing in this case). If we do decide to leave this in place, it might help to have a comment describing that as of the changes at this point, `__policy_can_access()` in BPF does not respect it.

ianvernon replied:
No, the logic skips this wildcard generate if the L4 Port is 0; we explicitly do not want to generate BPF entries with identity = 0 and port = 0 because this is not respected by the datapath. This is because we don't handle wildcarding at L3 in the datapath with a value of zero for the port. Did you read this as `if allows all at L3 and port = 0` instead of `if allows all at l3 and port != 0`? Your  statement leads me to think that you did. I can add a comment indicating why we aren't generating this key here.

joestringer replied:
Oh, weird, you've astutely noticed that I just completely mixed up the previous code vs. the new code, I read the diff as the opposite. Looking again this looks fine. Apologies for the noise!

joestringer reviewed (changes requested):
Apologies for the delay, we discussed offline and I thought I'd already published the review. I didn't closely inspect the duplicated policy iteration bits in ingress/egress since there's no direct diff, would be nice to tidy that up. As discussed offline I have a proposal to merge the ingress/egress logic paths which I can follow up on at a later date.

Commit:
policy: insert wildcard selector for L4 rules which allow all at L3

An empty slice of EndpointSelector within a rule has different meaning with
respect to policy at L3 when a rule is L3-only, or has L4 policy as well; an
empty slice of selectors when a rule is L3-only indicates that nothing is
allowed, while an empty slice of selectors when a rule has port-based policy
indicates that all L3 traffic is allowed at that port. Update the slice of
selectors being analyzed for L4 policy to contain the wildcard EndpointSelector
if a rule contains L4 policy, but has an empty slice of EndpointSelector to
account for this. This canonicalizes the behavior of an EndpointSelectorSlice to
mean that empty denies all at L3, while containing any selector allows all
traffic matching that selector.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: fix incorrect comments for function descriptions

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: move function applying on rule to rule.go

rules.go is for functions with ruleSlice as a receiver, so keep the code
organized.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
polcy: move calls to `selectRule` out of requirements analysis helper functions

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: rename functions which analyze ToEndpoints and FromEndpoints

This more accurately reflects what the functions do for better code readability.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: refactor canReach{Ingress,Egress} to use helper functions

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: add comment explaining why we can't generate wildcard L3 and wildcard L4 policy keys

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/policy/l4_test.go:
```diff
@@ -75,7 +75,7 @@ func (s *PolicyTestSuite) TestCreateL4Filter(c *C) {
 		eps := []api.EndpointSelector{selector}
 		// Regardless of ingress/egress, we should end up with
-		// a single L7 rule whether the selector is wildcarded
+		// a single L7 rule whether the selector is wildcardeds
```
What's wildcardeds? :)

ianvernon replied:
Fixed.

joestringer commented on pkg/policy/rule.go:
```diff
@@ -217,7 +214,28 @@ func mergeL4Ingress(ctx *SearchContext, rule api.IngressRule, ruleLabels labels.
+	)
+
+	// L3-only rule without requirements.
+	if len(rule.ToPorts) == 0 && len(rule.FromRequires) == 0 {
```
nit on function naming: this now makes `mergeL4Ingress()` include L3 which is a bit misleading. Same for egress side.

ianvernon replied:
Fixed - updated a bunch of other functions which included 'L4' in them as well. I specified port / proto as the thing being 'merged' since L3-only equates to '0/ANY' as a 'port-protocol' tuple. LMK if you think another naming convention is better.

joestringer commented on pkg/policy/rule.go:
```diff
@@ -410,6 +427,15 @@ func (r *rule) canReachIngress(ctx *SearchContext, state *traceState) api.Decisi
+
+	// At this point we only know whether anything is not denied by requirements.
+	// Now we need to determine whether rule allows traffic based off of
+	// SearchContext.
```
This comment is fine, but AFAICT the code immediately below was already run inside `r.meetsRequirementsIngress()`. I think the comment is trying to describe what the code below this block will do in `return r.canReachFromEndpoints()`. In other words, the iteration block immediately below this comment that ranges over the `r.FromRequires` in each `r.Ingress()` is duplicate and can be removed.

ianvernon replied:
Good catch - I think my brain remembered to delete it on the egress side, and thought it had done it on the ingress side too. Boilerplate =(

joestringer reviewed (changes requested):
Just a few remaining minor nits. Looks like the refactor was good for the egress side but unintentionally left an extra requirements iteration in on the ingress side. (Hopefully no brain farts this time :-) )

Commit:
policy: remove duplicate requirements check on Ingress

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: fix typo in comment

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: rename functions to reflect that L3-only policy is also generated

Signed-off by: Ian Vernon <ian@cilium.dev>
